core/domain/exp_services.py

```diff
@@ -586,6 +588,270 @@ def apply_change_list(exploration_id, change_list):
+            effective_old_to_new_state_names[old_state_name] = new_state_name
+    for old_state_name, new_state_name in (
+        effective_old_to_new_state_names.items()):
+        states_version_history[new_state_name] = (
+            state_domain.StateVersionHistory(
+                prev_version, old_state_name, committer_id))
+        del states_version_history[old_state_name]
+
+    # The following list includes states which exist in both the old states
+    # and new states and were not renamed.
+    states_which_were_not_renamed = []
+    for state_name in old_states:
+        if (
+            not(state_name in exp_versions_diff.deleted_state_names) and
+            not(state_name in effective_old_to_new_state_names)
+        ):
+            states_which_were_not_renamed.append(state_name)
+
+    # We have dealt with state additions, deletions and renames.
+    # Now we deal with states which were present in both versions and
+    # underwent changes only through the command EDIT_STATE_PROPERTY.
+    # The following dict stores whether the properties of states present
+    # in states_which_were_not_renamed were changed using EDIT_STATE_PROPERTY.
+    state_property_changed_data = {
+        state_name: False
+        for state_name in states_which_were_not_renamed
+    }
+    for change in change_list:
+        if (
+            change.cmd == exp_domain.CMD_EDIT_STATE_PROPERTY and
+            change.property_name != (
+                exp_domain.STATE_PROPERTY_RECORDED_VOICEOVERS) and
+            change.property_name != (
+                exp_domain.STATE_PROPERTY_WRITTEN_TRANSLATIONS)
+        ):
+            state_name = change.state_name
+            if state_property_changed_data.get(state_name) is False:
+                state_property_changed_data[state_name] = True
+
+    for state_name, state_property_changed in (
+        state_property_changed_data.items()):
+        if state_property_changed:
+            old_state_dict = copy.deepcopy(old_states_dict[state_name])
+            new_state_dict = copy.deepcopy(new_states_dict[state_name])
+
+            # Deleting the attributes written_translations and
+            # recorded_voiceovers because we don't want to consider diff
+            # in these properties as they are related to translations.
+            del old_state_dict['written_translations']
+            del old_state_dict['recorded_voiceovers']
+            del new_state_dict['written_translations']
+            del new_state_dict['recorded_voiceovers']
+
+            # The purpose of checking the diff_dict between the two state
+            # dicts ensure that we do not change the version history of that
+            # particular state if the overall changes (by EDIT_STATE_PROPERTY)
+            # get cancelled by each other and there is no 'net change'.
+            diff_dict = deepdiff.DeepDiff(
+                old_states_dict[state_name], new_states_dict[state_name])
+            if diff_dict:
```

Comment:
```suggestion
            if diff_dict is not None:
```
Or whatever the expected return is for no diff

Reply:
Done.